src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala

```diff
@@ -234,6 +244,18 @@ object DockerPlugin extends AutoPlugin {
       stage := Stager.stage(Docker.name)(streams.value, stagingDirectory.value, mappings.value),
       stage := (stage dependsOn dockerGenerateConfig).value,
       stagingDirectory := (target in Docker).value / "stage",
+      dockerLayerMappings := {
+        val dockerGroups = dockerLayerGrouping.value
+        def buildDockerPaths(seq: Seq[(File, String)], prefix: String) =
+          for {
+            (file, path) <- seq
+            layerIdx = dockerGroups(path)
+            pathInDocker = s"$prefix/$layerIdx/$path"
+          } yield (layerIdx, file, pathInDocker)
+
+        buildDockerPaths((mappings in Universal).value, (defaultLinuxInstallLocation in Docker).value + "/") ++
+          buildDockerPaths(dockerPackageMappings.value, "")
```

Comment:
It is a bit silly that we group `Universal / mappings` by their relative (to `defaultLinuxInstallLocation`) resulting paths, and `dockerPackageMappings` by absolute resulting paths, **using the same function**. Additionally, we mudden `Docker / mappings` with our layering strategy, which is not ideal. To be honest, I hadn't thought about `dockerPackageMappings` before this, and had assumed we'd only be grouping relative mappings. So it is partially my fault we've ended up here.

I propose we move the layering mechanism to **after** `Docker / mappings` in the build process, and make everything work with absolute resulting paths, as follows.

1. Keep `Docker / mappings` as it was before this PR. That is, the resulting paths (`._2` in the tuples) will still be absolute paths in the resulting Docker image, with no layering noise. This will allow us to avoid breaking anything that relies on `Docker / mappings` shape - e.g. the `dockerAdditionalPermissions` task.
1. `dockerLayerGrouping` should only work with absolute paths in the resulting Docker image. This would make writing it by hand even less convenient because of the `defaultLinuxInstallation` prefix (more on dealing with that later), but it should keep the abstraction cleaner.
1. Build `dockerLayerMappings` from `Docker / mappings` and `defaultLinuxInstallation`. The resulting paths (`._3` in the tuples) here will be absolute paths in the resulting Docker image, with no layering noise.
1. Modify `Docker / stage`. Instead of using `Docker / mappings` here, use `dockerLayerMappings`, and add layer prefixes to its resulting paths.
1. In `dockerCommands` copy each layer from its staging subdirectory with a separate command. The set of available layers can be extracted from `dockerLayerMappings`.
1. To simplify writing `dockerLayerGrouping`, introduce a helper function like this (the name is not final):

    ```scala
    def layerGroupingByStringPrefix(pairs: (String, Int)): String => Int = (s: String) =>
      pairs.find(p => s.startsWith(p._1)).map(_._2).getOrElse(0)
    ```

    which would allow us to write the default `dockerLayerGrouping` like this:

    ```scala
    dockerLayerGrouping := {
      val base = (defaultLinuxInstallLocation in Docker).value
      layerGroupingByStringPrefix(
        s"$base/lib/${organization.value}" -> 2,
        s"$base/bin/" -> 1
      )
    }
    ```

Reply:
Ok, will hack this tomorrow. I actually tried similar approach, but the grouping  function was less simple when I need to care about `defaultLinuxInstallLocation` prefix.

Reply:
When I follow this points I'll end up with staging directory structure like `target/docker/stage/0/opt/the-project/lib`. Notice `0` before `/opt` directory.
However first COPY command in the plugin assumes we copy everything from first directory of `defaultLinuxInstallLocation in Docker` which is /opt.
```
WORKDIR /opt/the-project
COPY opt /opt
```
[This](https://github.com/sbt/sbt-native-packager/blob/master/src/main/scala/com/typesafe/sbt/packager/docker/DockerPlugin.scala#L326) means that the very first directory of linux installation it the only source of all files. Even files from `dockerPackageMappings` need to be mapped to /opt directory. This means I should do special handling of the first directory, and add layer_id directories directly after it.

Reply:
> However first COPY command in the plugin assumes we copy everything from first directory of defaultLinuxInstallLocation in Docker which is /opt.

If we want to have separate layers, we will have to replace this with `COPY` commands for each layer. That's the main point of this PR. So we would have something like this:

```Dockerfile
WORKDIR /opt/the-project
COPY 0/* /
COPY 1/* /
...
```

> This means I should do special handling of the first directory, and add layer_id directories directly after it.

Not sure why we would need to do any extra handling. All the files we want copied should be in the layer directories (except for the `Dockerfile`), so we don't need co copy `/opt` separately. I'm probably missing something, since I'm not very familiar with the subject matter, and because it's getting late here - could you please elaborate as to why just copying layer directories as shown above is not sufficient?

Reply:
First part of answer is enough to solve this now.